client/components/edit/Breadcrumbs.tsx

```diff
@@ -0,0 +1,147 @@
+import styles from "./Breadcrumbs.module.css";
+import React, { useEffect, useState } from "react";
+import PropTypes from "prop-types";
+import { useFetchContext } from "../../context/FetchContext";
+import { apiUrl } from "../../util/routes";
+import Link from "next/link";
+
+interface TreeNode {
+    pid: string;
+    title: string;
+    parents: Array<TreeNode>;
+}
+
+interface TreeData {
+    topNodes: Array<string>;
+    records: Record<string, TreeNode>;
+    childLookups: Record<string, Array<string>>;
+}
+
+interface BreadcrumbTrail {
+    pid: string;
+    path: Array<TreeNode>;
+}
+
+const Breadcrumbs = ({ pid = null }) => {
+    const {
+        action: { fetchJSON },
+    } = useFetchContext();
+    const [treeData, setTreeData] = useState([]);
+
+    /**
+     * Analyze the raw breadcrumb data, identifying top-level nodes (i.e. nodes with
+     * no parents) and creating lookup tables for children. This makes it easier to
+     * render breadcrumbs from left to right and to find all relevant trails.
+     */
+    function processBreadcrumbData(data: TreeNode): TreeData {
+        const queue: Array<TreeNode> = [data];
+        const topNodes: Array<string> = [];
+        const childLookups: Record<string, Array<string>> = {};
+        const records: Record<string, TreeNode> = {};
+        while (queue.length > 0) {
+            const current = queue.shift();
+            if (!current) {
+                break;
+            }
+            if (current.parents.length === 0) {
+                topNodes.push(current.pid);
+            }
+            records[current.pid] = { pid: current.pid, title: current.title, parents: [] };
+            current.parents.forEach((parent) => {
+                queue.push(parent);
+                if (typeof childLookups[parent.pid] === "undefined") {
+                    childLookups[parent.pid] = [current.pid];
+                } else {
+                    childLookups[parent.pid].push(current.pid);
+                }
+            });
+        }
+        return {
+            topNodes: [...new Set(topNodes)], // deduplicate
+            records,
+            childLookups,
+        };
+    }
+
+    useEffect(() => {
+        async function loadData() {
+            let data: TreeData = {
+                topNodes: [],
+                childLookups: {},
+                records: {}
+            };
+            const url = apiUrl + "/edit/object/parents/" + encodeURIComponent(pid);
+            try {
+                data = processBreadcrumbData(await fetchJSON(url));
+            } catch (e) {
+                console.error("Problem fetching breadcrumb data from " + url);
+            }
+            setTreeData(data);
+        }
+        loadData();
+    }, []);
+
+    function generateBreadcrumbTrails(treeData: TreeData, pid: string) {
+        // BFS from top (root id) to target pid
+        const queue: Array<BreadcrumbTrail> = [];
```

Comment:
U can shorten using treeData.map.

Reply:
Good point! I'll fix that in the near future...

Reply:
Fixed! Thanks again for the suggestion. Let me know if you have any other feedback before we merge this.